File: src/algorithms/bfs.js
export const bfs = (graph, startNodeId, endNodeId) => {
  const queue = [startNodeId];
  const visited = new Set([startNodeId]);
  const previous = {};
  const visitedOrder = [];
  const distances = {}; // For consistent return format

  Object.keys(graph).forEach(node => {
    previous[node] = null;
    distances[node] = Infinity;
  });
  distances[startNodeId] = 0;
  let found = false;
  while (queue.length > 0) {
    const current = queue.shift();
    visitedOrder.push(current);
    if (current === endNodeId) {
      found = true;
      break;
    }
    const neighbors = graph[current].neighbors;
    for (const neighborId in neighbors) {
      if (!visited.has(neighborId)) {
        visited.add(neighborId);
        previous[neighborId] = current;
        distances[neighborId] = distances[current] + neighbors[neighborId]; // Accumulate distance for accuracy
        queue.push(neighborId);
      }
    }
  }

  // Reconstruct path
  const path = [];
  if (found) {
    let u = endNodeId;
    while (u !== null) {
      path.unshift(u);
      u = previous[u];
    }
  }
  return {
    visitedOrder,
    path,
    distance: distances[endNodeId]
  };
};

File: java_algorithms/BFS.java
package java_algorithms;

import java.util.*;

public class BFS {

    public static Result findPath(Map<String, Node> graph, String startNodeId, String endNodeId) {
        Queue<String> queue = new LinkedList<>();
        queue.add(startNodeId);
        
        Set<String> visited = new HashSet<>();
        visited.add(startNodeId);
        
        Map<String, String> previous = new HashMap<>();
        List<String> visitedOrder = new ArrayList<>();
        Map<String, Double> distances = new HashMap<>();

        for (String node : graph.keySet()) {
            previous.put(node, null);
            distances.put(node, Double.POSITIVE_INFINITY);
        }
        distances.put(startNodeId, 0.0);

        boolean found = false;
        while (!queue.isEmpty()) {
            String current = queue.poll();
            visitedOrder.add(current);

            if (current.equals(endNodeId)) {
                found = true;
                break;
            }

            Map<String, Double> neighbors = graph.get(current).neighbors;
            for (Map.Entry<String, Double> entry : neighbors.entrySet()) {
                String neighborId = entry.getKey();
                double weight = entry.getValue();

                if (!visited.contains(neighborId)) {
                    visited.add(neighborId);
                    previous.put(neighborId, current);
                    distances.put(neighborId, distances.get(current) + weight);
                    queue.add(neighborId);
                }
            }
        }

        // Reconstruct path
        List<String> path = new ArrayList<>();
        if (found) {
            String u = endNodeId;
            while (u != null) {
                path.add(0, u);
                u = previous.get(u);
            }
        }

        return new Result(visitedOrder, path, distances.get(endNodeId));
    }
}
